File: 2791_count_paths_that_can_form_a_palindrome_in_a_tree/Solution.cs
// LeetCode 2791 - Count Paths That Can Form a Palindrome in a Tree
// https://leetcode.com/problems/count-paths-that-can-form-a-palindrome-in-a-tree/

using System.Collections.Generic;

public class Solution {
    public long CountPalindromePaths(IList<int> parent, string s) {
        int n = parent.Count;
        var g = new List<int>[n];
        for (int i = 0; i < n; i++) g[i] = new List<int>();
        for (int i = 1; i < n; i++) g[parent[i]].Add(i);
        var freq = new Dictionary<int, int>();
        freq[0] = 1;
        long ans = 0;
        void Dfs(int u, int mask) {
            foreach (int v in g[u]) {
                int nm = mask ^ (1 << (s[v] - 'a'));
                if (freq.TryGetValue(nm, out int c0)) ans += c0;
                for (int b = 0; b < 26; b++) {
                    int key = nm ^ (1 << b);
                    if (freq.TryGetValue(key, out int c)) ans += c;
                }
                if (!freq.ContainsKey(nm)) freq[nm] = 0;
                freq[nm]++;
                Dfs(v, nm);
            }
        }
        Dfs(0, 0);
        return ans;
    }
}

File: 2791_count_paths_that_can_form_a_palindrome_in_a_tree/solution.rs
// LeetCode 2791 - Count Paths That Can Form a Palindrome in a Tree
// https://leetcode.com/problems/count-paths-that-can-form-a-palindrome-in-a-tree/

use std::collections::HashMap;

impl Solution {
    pub fn count_palindrome_paths(parent: Vec<i32>, s: String) -> i64 {
        let n = parent.len();
        let mut g = vec![Vec::new(); n];
        for i in 1..n {
            g[parent[i] as usize].push(i);
        }
        let bytes = s.as_bytes();
        let mut freq = HashMap::new();
        freq.insert(0i32, 1i64);
        let mut ans = 0i64;
        fn dfs(
            u: usize,
            mask: i32,
            g: &[Vec<usize>],
            bytes: &[u8],
            freq: &mut HashMap<i32, i64>,
            ans: &mut i64,
        ) {
            for &v in &g[u] {
                let nm = mask ^ (1 << (bytes[v] - b'a'));
                *ans += *freq.get(&nm).unwrap_or(&0);
                for b in 0..26 {
                    *ans += *freq.get(&(nm ^ (1 << b))).unwrap_or(&0);
                }
                *freq.entry(nm).or_insert(0) += 1;
                dfs(v, nm, g, bytes, freq, ans);
            }
        }
        dfs(0, 0, &g, bytes, &mut freq, &mut ans);
        ans
    }
}
